sycl_runtime/lib/sycl_runtime.cpp

```diff
@@ -0,0 +1,260 @@
+  }
+
+  bool operator!=(const ParamDesc &rhs) const { return !(*this == rhs); }
+};
+
+template <typename T> size_t countUntil(T *ptr, T &&elem) {
+  assert(ptr);
+  auto curr = ptr;
+  while (*curr != elem) {
+    ++curr;
+  }
+  return static_cast<size_t>(curr - ptr);
+}
+
+sycl::device get_default_device() {
+  auto platform_list = sycl::platform::get_platforms();
+  for (const auto &platform : platform_list) {
+    auto platform_name = platform.get_info<sycl::info::platform::name>();
+    bool is_level_zero = platform_name.find("Level-Zero") != std::string::npos;
+    if (!is_level_zero)
+      continue;
+
+    return platform.get_devices()[0];
+  }
+}
+
+struct GPU_SYCL_QUEUE {
+
+  sycl::device sycl_device_;
+  sycl::context sycl_context_;
+  sycl::queue sycl_queue_;
+
+  GPU_SYCL_QUEUE() {
+
+    sycl_device_ = get_default_device();
+    sycl_context_ = sycl::context(sycl_device_);
+    sycl_queue_ = sycl::queue(sycl_context_, sycl_device_);
+  }
+
+  GPU_SYCL_QUEUE(sycl::device *device, sycl::context *context) {
+    sycl_device_ = *device;
+    sycl_context_ = *context;
+    sycl_queue_ = sycl::queue(sycl_context_, sycl_device_);
+  }
+  GPU_SYCL_QUEUE(sycl::device *device) {
+
+    sycl_device_ = *device;
+    sycl_context_ = sycl::context(sycl_device_);
+    sycl_queue_ = sycl::queue(sycl_context_, sycl_device_);
+  }
+
+  GPU_SYCL_QUEUE(sycl::context *context) {
+
+    sycl_device_ = get_default_device();
+    sycl_context_ = *context;
+    sycl_queue_ = sycl::queue(sycl_context_, sycl_device_);
+  }
+
+}; // end of GPU_SYCL_QUEUE
+
+void *alloc_device_memory(void *queue, size_t size, size_t alignment,
+                          bool is_shared) {
+  void *mem_ptr = nullptr;
+  if (is_shared) {
+    mem_ptr = sycl::aligned_alloc_shared(
+        alignment, size, static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_);
+  } else {
+    mem_ptr = sycl::aligned_alloc_device(
+        alignment, size, static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_);
+  }
+  return mem_ptr;
+}
+
+void dealloc_device_memory(void *queue, void *ptr) {
+  sycl::free(ptr, static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_);
+}
+
+ze_module_handle_t loadModule(void *queue, const void *data, size_t dataSize) {
+  assert(data);
+  auto sycl_queue = static_cast<GPU_SYCL_QUEUE *>(queue)->sycl_queue_;
+  ze_module_handle_t ze_module;
+  ze_module_desc_t desc = {ZE_STRUCTURE_TYPE_MODULE_DESC,
+                           nullptr,
+                           ZE_MODULE_FORMAT_IL_SPIRV,
+                           dataSize,
+                           (const uint8_t *)data,
+                           nullptr,
+                           nullptr};
+  auto ze_device = sycl::get_native<sycl::backend::ext_oneapi_level_zero>(
+      sycl_queue.get_device());
+  auto ze_ctx = sycl::get_native<sycl::backend::ext_oneapi_level_zero>(
+      sycl_queue.get_context());
+  L0_SAFE_CALL(zeModuleCreate(ze_ctx, ze_device, &desc, &ze_module, nullptr));
```

Comment:
Like we discussed, it is too much overhead by creating module and kernel compilation every time.  We need to get rid of this overhead as soon as possible. But We can take it as next step. Need a good design for this.